examples/mxnet/gcmc/train_es_minibatch.py

```diff
@@ -0,0 +1,334 @@
+                                    os.path.join(args.save_dir, 'test_loss%d.csv' % args.save_id))
+
+    ### declare the loss information
+    best_valid_rmse = np.inf
+    no_better_valid = 0
+    best_iter = -1
+
+    enc_graph = dataset.train_enc_graph
+    nd_possible_rating_values = mx.nd.array(dataset.possible_rating_values, ctx=args.ctx, dtype=np.float32)
+    g_user_fea = mx.nd.zeros((dataset.num_user,))
+    g_movie_fea = mx.nd.zeros((dataset.num_movie,))
+    train_truths = dataset.train_truths
+    train_labels = dataset.train_labels
+
+    print("Start training ...")
+    dur = []
+
+    for iter_idx in range(1, args.train_max_iter):
+        if iter_idx > 3:
+            t0 = time.time()
+
+        num_edges = dataset.train_truths.shape[0]
+        seed = mx.nd.arange(num_edges, dtype='int64')
+        edges = mx.nd.shuffle(seed)
+        # each iteration will go through all edges
+        for sample_idx in range(0, (num_edges + args.minibatch_size - 1) // args.minibatch_size):
+            edge_ids = edges[sample_idx * args.minibatch_size: (sample_idx + 1) * args.minibatch_size if (sample_idx + 1) * args.minibatch_size < num_edges else num_edges]
+            head_ids, tail_ids = dataset.train_dec_graph.find_edges(edge_ids.asnumpy())
+
+            head_subgraphs = {}
+            tail_subgraphs = {}
+            head_node_ids = np.unique(head_ids.asnumpy())
+            tail_node_ids = np.unique(tail_ids.asnumpy())
+            for i, _ in enumerate(args.rating_vals):
+                t = enc_graph.canonical_etypes[i * 2]
+                rev_t = enc_graph.canonical_etypes[i * 2 + 1]
+
+                head_in_edges = enc_graph.in_edges(head_node_ids, 'eid', etype=rev_t)
+                tail_in_edges = enc_graph.in_edges(tail_node_ids, 'eid', etype=t)
+
+                if head_in_edges.shape[0] > 0:
+                    head_subgraphs[rev_t] = head_in_edges
+
+                if tail_in_edges.shape[0] > 0:
+                    tail_subgraphs[t] = tail_in_edges
+
+            head_subgraph = enc_graph.edge_subgraph(head_subgraphs, preserve_nodes=True)
+            tail_subgraph = enc_graph.edge_subgraph(tail_subgraphs, preserve_nodes=True)
+            edge_ids = edge_ids.as_in_context(args.ctx)
+            true_relation_ratings = train_truths[edge_ids]
+            true_relation_labels = train_labels[edge_ids]
+
+            head_NID = head_subgraph.nodes['user'].data[dgl.NID]
+            tail_NID = tail_subgraph.nodes['movie'].data[dgl.NID]
+
+            g_user_fea[head_NID] = mx.nd.arange(head_NID.shape[0], dtype='int32')
+            g_movie_fea[tail_NID] = mx.nd.arange(tail_NID.shape[0], dtype='int32')
+
+            true_head_idx = g_user_fea[head_ids].as_in_context(args.ctx)
+            true_tail_idx = g_movie_fea[tail_ids].as_in_context(args.ctx)
+
+            with mx.autograd.record():
+                pred_ratings = net(head_subgraph, tail_subgraph,
+                                   true_head_idx, true_tail_idx)
```

Comment:
I see why you need a new model class. However, generally we want to use the same model definition regardless it is a mini-batch training or full graph training.
If I understand the code correctly, the main issue of mini-batch sampling is that if the sampled edges in the encoder graph are not incident to *any* edges in the decoder graph, the whole encoder-decoder pipeline breaks. As a result, you first sample edges from the decoder graph, then retrieve the user/item nodes touched by those edges, and finally extract the encoder subgraph that contributes to the user/item nodes.
However, I feel we could still tweak the model interface to be exactly the same as the old one. Is there any specific reason why this cannot be done?

Reply:
I am wondering if we can use graph.pull instead of upate_all in mini-batch. which donot need to generate subgraph.